willow/registry.py

```diff
@@ -26,15 +33,29 @@ class UnroutableOperationError(LookupError):
 
 class WillowRegistry(object):
     def __init__(self):
+        self._used_operation_names = set()
         self._registered_image_classes = set()
         self._unavailable_image_classes = dict()
         self._registered_operations = defaultdict(dict)
+        self._registered_generic_operations = dict()
         self._registered_converters = dict()
         self._registered_converter_costs = dict()
 
     def register_operation(self, image_class, operation_name, func):
+        if operation_name in self._registered_generic_operations:
+            return OperationNameConflict("A generic operation already exists with the name '{0}'".format(operation_name))
+
         self._registered_operations[image_class][operation_name] = func
 
+    def register_generic_operation(self, dependencies, operation_name, func):
```

Comment:
an "operation_name, func, dependencies" order makes more sense to me.

I realise you are modelling this on `register_operation` above, but the distinction is that  the image_class is in the right place as it defines where the operation takes place...